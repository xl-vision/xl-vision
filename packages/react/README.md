# `@xl-vision/react`

> TODO: description

