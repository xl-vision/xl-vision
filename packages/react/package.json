{
  "name": "@xl-vision/react",
  "version": "0.0.0",
  "description": "xl-vision react components.",
  "author": "Rhys Xia <xrs4433@outlook.com>",
  "license": "MIT",
  "main": "legacy/index.js",
  "module": "modern/index.js",
  "types": "legacy/index.d.ts",
  "files": [
    "legacy",
    "modern"
  ],
  "keywords": [
    "xl-vision",
    "react",
    "styles"
  ],
  "homepage": "https://github.com/xl-vision/xl-vision-next/tree/master/packages/react#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/xl-vision/xl-vision-next.git"
  },
  "scripts": {
    "build:legacy": "node ../../scripts/build legacy",
    "build:modern": "node ../../scripts/build modern",
    "build": "npm run build:legacy && npm run build:modern"
  },
  "bugs": {
    "url": "https://github.com/xl-vision/xl-vision-next/issues"
  },
  "dependencies": {
    "@babel/runtime": "^7.12.5",
    "prop-types": "^15.7.2"
  },
  "devDependencies": {
    "@types/react": "^16.8.0",
    "@types/react-dom": "^16.8.0",
    "@xl-vision/styled-engine": "^0.0.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  },
  "peerDependencies": {
    "@types/react": "^16.8.0 || ^17.0.0",
    "@xl-vision/styled-engine": "^0.0.0",
    "react": "^16.8.0 || ^17.0.0",
    "react-dom": "^16.8.0 || ^17.0.0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    }
  },
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  }
}
