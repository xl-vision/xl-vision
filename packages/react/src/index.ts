export { styled, createGlobalStyles } from './styles';
export { default as ThemeProvider } from './ThemeProvider';
export { default as Transition } from './Transition';
export { default as CSSTransition } from './CSSTransition';
export { default as CollapseTransition } from './CollapseTransition';
export { default as TransitionGroup } from './TransitionGroup';
