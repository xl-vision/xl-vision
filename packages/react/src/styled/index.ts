export { styled as default } from '../styles';
