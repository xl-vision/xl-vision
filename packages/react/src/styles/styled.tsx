import {
  CSSObject,
  ExtractProps,
  FilteringStyledOptions,
  FunctionInterpolation,
  Interpolation,
  StyledComponent,
} from '@xl-vision/styled-engine-types';
import innerStyled from '@xl-vision/styled-engine';
import React from 'react';
import { Theme } from './createTheme';
import ThemeContext from './ThemeContext';

const styled = <
  Tag extends keyof JSX.IntrinsicElements | React.ComponentType<React.ComponentProps<Tag>>,
  ForwardedProps extends keyof ExtractProps<Tag> = keyof ExtractProps<Tag>,
  StyleProps extends { theme: Theme } = { theme: Theme }
>(
  tag: Tag,
  options?: FilteringStyledOptions<ExtractProps<Tag>, ForwardedProps>,
) => {
  const defaultCreateStyledComponent = innerStyled<Tag, ForwardedProps, StyleProps>(tag, options);

  const overrideCreateStyledComponent = <AdditionalProps extends object = {}>(
    first:
      | TemplateStringsArray
      | CSSObject
      | FunctionInterpolation<
          Pick<ExtractProps<Tag>, ForwardedProps> & StyleProps & AdditionalProps
        >,
    ...styles: Array<
      Interpolation<Pick<ExtractProps<Tag>, ForwardedProps> & StyleProps & AdditionalProps>
    >
  ) => {
    const DefaultComponent = defaultCreateStyledComponent(first, ...styles);

    const Cmp: StyledComponent<
      Pick<ExtractProps<Tag>, ForwardedProps>,
      Omit<StyleProps, 'theme'> & { theme?: Theme }
    > = (props) => {
      // eslint-disable-next-line react/prop-types
      const { theme: themeProp, ...others } = props;

      const defaultTheme = React.useContext(ThemeContext);

      const theme = themeProp || defaultTheme;
      // @ts-ignore
      return <DefaultComponent {...others} theme={theme} />;
    };

    return Cmp;
  };

  return overrideCreateStyledComponent;
};

export default styled;
