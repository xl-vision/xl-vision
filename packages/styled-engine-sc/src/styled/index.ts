import { Styled } from '@xl-vision/styled-engine-types';
import scStyled from 'styled-components';

const styled: Styled = (tag, options) => {
  if (options) {
    const { shouldForwardProp, label } = options;
    return scStyled(tag).withConfig({
      shouldForwardProp,
      displayName: label,
    }) as ReturnType<Styled>;
  }
  return scStyled(tag) as ReturnType<Styled>;
};

export default styled;
