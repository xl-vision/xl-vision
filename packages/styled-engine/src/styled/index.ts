import { Styled } from '@xl-vision/styled-engine-types';
import scStyled from 'styled-components';

const styled: Styled = (tag, options) => {
  if (options) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return scStyled(tag).withConfig(options) as any;
  }
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return scStyled(tag) as any;
};

export default styled;
