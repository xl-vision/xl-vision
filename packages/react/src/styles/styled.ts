import {
  styled as innerStyled,
  CSSObject,
  ExtractProps,
  FunctionInterpolation,
  Interpolation,
  ShouldForwardProp,
} from '@xl-vision/styled-engine';
import { env } from '@xl-vision/utils';
import { ComponentProps, ComponentType } from 'react';
import { Theme } from '../ThemeProvider/createTheme';
import { Style } from '../ThemeProvider/overrideStyles';
import applyTheme from './applyTheme';

export type XlOptions = {
  name?: string;
  slot?: string;
};

const shouldForwardProp = (prop: PropertyKey) => prop !== 'theme' && prop !== 'styleProps';

const middleline = (str: string) => {
  const separator = '-';
  const split = /(?=[A-Z])/;

  return str.split(split).join(separator).toLowerCase();
};

const styled = <
  Tag extends keyof JSX.IntrinsicElements | ComponentType<ComponentProps<Tag>>,
  ForwardedProps extends keyof ExtractProps<Tag> = keyof ExtractProps<Tag>,
>(
  tag: Tag,
  options?: XlOptions,
) => {
  const { name, slot = 'Root' } = options || {};

  let displayName = '';
  let prefix = '';

  if (name) {
    displayName = name + slot;
    if (slot === 'Root') {
      prefix = middleline(name);
    }
  }

  const defaultCreateStyledComponent = innerStyled<Tag, ForwardedProps>(tag, {
    shouldForwardProp: shouldForwardProp as ShouldForwardProp<ForwardedProps>,
    ...(env.isDevelopment && { prefix: prefix || undefined }),
  });

  const overrideCreateStyledComponent = <
    S extends {} | undefined = undefined,
    P extends Pick<ExtractProps<Tag>, ForwardedProps> = Pick<ExtractProps<Tag>, ForwardedProps>,
    E = S extends undefined ? { theme: Theme } : { styleProps: S; theme: Theme },
    V = Omit<E, 'theme'> & { theme?: Theme },
  >(
    first: TemplateStringsArray | CSSObject | FunctionInterpolation<P, E>,
    ...styles: Array<Interpolation<P, E>>
  ) => {
    const applyOverrideStyle = (props: P & E & { theme: Theme }) => {
      const { theme } = props;
      if (!name || !slot) {
        return;
      }
      const overrideStyles = theme.overrideStyles as Record<string, Record<string, Style<S, P>>>;

      const overrideStyle = overrideStyles[name];

      if (!overrideStyle) {
        return;
      }
      const overrideSlotStyle = overrideStyle[slot];
      if (!overrideSlotStyle) {
        return;
      }
      if (typeof overrideSlotStyle === 'function') {
        // TODO [2022-04-01]: type fix
        return (overrideSlotStyle as unknown as FunctionInterpolation<P, E>)(props);
      }
      return overrideSlotStyle;
    };

    const newStyles = [...styles, applyOverrideStyle].map(applyTheme);

    let newFirst: any = first;

    const numOfCustomFnsApplied = newStyles.length - styles.length;

    if (Array.isArray(newFirst) && numOfCustomFnsApplied > 0) {
      const newFirstArray = newFirst as unknown as TemplateStringsArray;
      const placeholders = new Array<string>(numOfCustomFnsApplied).fill('');
      const raw = [...newFirstArray.raw, ...placeholders];
      newFirst = [...newFirstArray, ...placeholders];
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      newFirst.raw = raw;
    } else if (typeof newFirst === 'function') {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      newFirst = applyTheme(newFirst);
    }

    const DefaultComponent = defaultCreateStyledComponent<P & V>(
      newFirst as TemplateStringsArray | CSSObject | FunctionInterpolation<P, V>,
      ...newStyles,
    );

    if (!env.isProduction) {
      DefaultComponent.displayName = displayName;
    }

    return DefaultComponent;
  };

  return overrideCreateStyledComponent;
};

export default styled;
