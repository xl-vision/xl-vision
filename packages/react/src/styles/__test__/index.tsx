import { render } from '@testing-library/react';
import { useMemo } from 'react';
import { ThemeProvider, ThemeInput } from '@xl-vision/react';
import styled from '../styled';

const Button = styled('button', {
  name: 'DemoButton',
  slot: 'Root',
})(() => {
  return {
    color: 'red',
  };
});

describe('styled', () => {
  it('test styled overrideStyles', () => {
    const Demo = ({ color }: { color?: string }) => {
      const theme = useMemo(() => {
        if (!color) {
          return {};
        }
        return {
          overrideStyles: {
            DemoButton: {
              Root: {
                color,
              },
            },
          },
        } as ThemeInput;
      }, [color]);
      return (
        <ThemeProvider {...theme}>
          <Button>click</Button>
        </ThemeProvider>
      );
    };

    const { container, rerender } = render(<Demo />);

    expect(container).toMatchSnapshot();

    rerender(<Demo color='blue' />);

    expect(container).toMatchSnapshot();
  });
});
