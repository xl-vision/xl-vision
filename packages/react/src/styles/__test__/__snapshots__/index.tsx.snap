// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled test styled overrideStyles 1`] = `
.c0 {
  color: red;
}

<div>
  <button
    class="c0 xl-demo-button"
  >
    click
  </button>
</div>
`;

exports[`styled test styled overrideStyles 2`] = `
.c0 {
  color: red;
  color: blue;
}

<div>
  <button
    class="c0 xl-demo-button"
  >
    click
  </button>
</div>
`;
